primap2/csg/_strategies/local_trends.py

```diff
@@ -0,0 +1,267 @@
+    boundary) non-NaN data point in :math:`\\textrm{ts}`. The trend value is calculated
+    using a linear trend of length `trend_length` or less data points if a time-series
+    does not cover the full period. By setting `min_trend_points` a minimal number of
+    points necessary for the trend calculation can be set. If less points are available a
+    :py:class:`StrategyUnableToProcess` error will be raised. This enables the user to
+    define a fallback strategy, e.g. single point matching.
+    TODO: for the case of gaps this leads to the situation that we can't use trends on
+    one side of the gap and single year matching as fallback on the other
+
+    By setting `trend_length` to 1 single year matching is used.
+
+    For gaps the left (:math:`t_{bl}`) and right (:math:`t_{br}`) end have to be considered.
+    The data is harmonized using
+
+    .. math::
+
+       \\textrm{fill_ts}_h(t) = \\textrm{fill_ts}(t) \\times
+       \\frac{a_l(t_{br}-t)+a_r(t-t_{bl})}{t_{br}-t_{bl}},
+
+    where
+
+    .. math::
+
+        a_l = \\textrm{fill_ts}_t(t_{bl}) / \\textrm{ts}_t(t_{bl}),
+
+    and
+
+    .. math::
+
+        a_r = \\textrm{fill_ts}_t(t_b) / \\textrm{ts}_t(t_b).
+
+    If only one of the ends of the gap has an overlap with :math:`\\textrm{fill_ts}(t)`,
+    we use the harmonization factor of this side for the whole gap (we treat the gap like
+    a boundary).
+
+    If there is no overlap in non-NaN data between :math:`\\textrm{ts}(t)` and
+    :math:`\\textrm{fill_ts}(t)` a :py:class:`StrategyUnableToProcess` error will be raised.
+
+    If ``allow_negative = False`` and the harmonized time-series :math:`\\textrm{fill_ts}_h(t)`
+    contains negative data a :py:class:`StrategyUnableToProcess` error will be raised.
+
+    Filling multiple gaps and boundaries with this function is scientifically questionable
+    as they will all use different scaling factors and thus don't use a consistent model to
+    harmonize one time-series :math:`\\textrm{fill_ts}(t)` to :math:`\\textrm{ts}(t)`.
+    Use with care.
+
+    Attributes
+    ----------
+    trend_length: int, default 1
+        Define the number of data point used to calculate trend values for the matching
+        point
+    min_trend_points: int, default 1
+        Minimal number of data points needed for trend calculation. Can't be larger than
+        trend_length obviously
+    allow_negative: bool, default False
+        Allow the filling time series to contain negative data initially.
+    """
+
+    # TODO: add fallback options so we can calculate a trend for one to also if the
+    #  other doesn't have sufficient information. similarly if we have several gaps to fill
+    #  we can use trends where possible, but not for all gaps if data is not sufficient
+    #  for all gaps
+    # fallback_option
+
+    fit_params: FitParameters = field()
+    allow_negative: bool = field()  # not implemented yet
```

Comment:
```suggestion
    fit_params: FitParameters
    allow_negative: bool # not implemented yet
```

it becomes a `field()` automatically, no need to specify it.

Also, should the `allow_negative` be removed while it is not implemented yet?

Reply:
I think ruff complained and that's why I added the field statement. But I'll try again, maybe I missed that combination.
Regarding `allow_negative`: It's implemented, just forgot to remove the comment

Reply:
I get `NameError: name 'fit_params' is not defined` without the field statement

Reply:
Ah, now I see why. you are later setting `@fit_params.default`, which only works if you `field()` it. Why is that necessary, why can't you just write

```suggestion
    fit_params: FitParameters = FitParameters(trend_length=10, min_trend_points=5, trend_length_unit="YS", fit_degree=1, fallback_degree=0)
    allow_negative: bool = False
```

and delete the `@fit_params.default` and `@allow_negative.default` functions?

Reply:
I don't have the exact error in mind, but I think it had something to do with making it a class or instance variable.

Reply:
The ruff complaint is:
`RUF009 Do not perform function call `FitParameters` in dataclass defaults`

Was quite a bit of work/research to fix it as the solution presented in the ruff docs doesn't work either for this case.  It involved me actually looking into a book (a proper physically existing book). But in the end the `attrs` docs helped if I remember correctly. At least I learned a bit about class and instance variables and that tools to make declarations easier sometimes make them more complicated.

Reply:
hm, okay, makes sense. Always a bit harder to read with the `field` stuff, but probably it's due to `FitParameters` being a class itself.